src/polycubed/src/controller.cpp

```diff
@@ -136,8 +141,19 @@ int controller_module_rx(struct __sk_buff *ctx) {
 
 	u16 in_port = md->port_id;
 	u16 module_index = md->module_index;
+    u32 flags = md->flags;
 
 	ctx->cb[0] = in_port << 16 | module_index;
+    ctx->cb[2] = flags;
+    if (module_index == 0xffff) {
+        pcn_log(ctx, LOG_INFO, "[tc-decapsulator]: NH is stack, flags: 0x%x", flags);
+        if (flags & MD_EGRESS_CONTEXT) {
+            return bpf_redirect(in_port, 0);
+        } else {
+            return 0;
```

Comment:
What is `return 0` for in this case?

Reply:
return 0 (TC_ACK_OK) means the packet will pass to Linux stack. For packet forwarded to slow path in ingress context should continue the stack process (routing, filter, egress BPF etc.) before it was sent out.

Reply:
It makes sense. There only one point I am not totally sure. This eBPF program is executed in the context of the `pcn_tc_cp` ingress hook, then the OS will think the packet is coming from that interface. Do you think this is a problem?

Reply:
The ingress interface is not determined by the interface hook where BPF program is executed (bpf_call can change the context at any time), the interface is determined by the interface driver that picking up the packet, which in our case pcn_tc_cp when packet re-injected to this interface by calling send_packet_to_cube(). But recall that the packet is originally arriving in another interface (with dest MAC to that interface) before it was sent to slow path, without modifying the dest MAC kernel wouldn't happy with that.

Reply:
I'm not totally sure I understood your comment.
Let's think about this case, a transparent cube is attached to eth0.
A packet arrives to `eth0`, it is sent to the slowpath and then the slow paths performs a `packet_out` to send it to the linux stack. With this architecture the packet will be passed to the linux stack from the `pcn_tc_cp` interface and not from `eth0`.  Is this is a problem besides the MAC address?